src/middleware/usbboot.rs

```diff
@@ -136,83 +112,20 @@ pub(crate) fn verify_one_device<T>(devices: &[T]) -> std::result::Result<(), Fla
     }
 }
 
-pub(crate) async fn get_device_path<I: IntoIterator<Item = &'static str>>(
-    allowed_vendors: I,
-) -> anyhow::Result<PathBuf, FlashingError> {
-    let mut contents = fs::read_dir("/dev/disk/by-id").await.map_err(|err| {
-        log::error!("Failed to list devices: {}", err);
-        FlashingError::IoError
-    })?;
-
-    let target_prefixes = allowed_vendors
-        .into_iter()
-        .map(|vendor| format!("usb-{}_", vendor))
-        .collect::<Vec<String>>();
-
-    let mut matching_devices = vec![];
-
-    while let Some(entry) = contents.next_entry().await.map_err(|err| {
-        log::warn!("Intermittent IO error while listing devices: {}", err);
-        FlashingError::IoError
-    })? {
-        let Ok(file_name) = entry.file_name().into_string() else {
-            continue;
-        };
-
-        for prefix in &target_prefixes {
-            if file_name.starts_with(prefix) {
-                matching_devices.push(file_name.clone());
-            }
-        }
-    }
-
-    // Exclude partitions, i.e. turns [ "x-part2", "x-part1", "x", "y-part2", "y-part1", "y" ]
-    // into ["x", "y"].
-    let unique_root_devices = matching_devices
-        .iter()
-        .filter(|this| {
-            !matching_devices
-                .iter()
-                .any(|other| this.starts_with(other) && *this != other)
-        })
-        .collect::<Vec<&String>>();
-
-    let symlink = match unique_root_devices[..] {
-        [] => {
-            log::error!("No supported devices found");
-            return Err(FlashingError::DeviceNotFound);
-        }
-        [device] => device.clone(),
-        _ => {
-            log::error!(
-                "Several supported devices found: found {}, expected 1",
-                unique_root_devices.len()
-            );
-            return Err(FlashingError::GpioError);
-        }
-    };
-
-    fs::canonicalize(format!("/dev/disk/by-id/{}", symlink))
-        .await
-        .map_err(|err| {
-            log::error!("Failed to read link: {}", err);
-            FlashingError::IoError
-        })
-}
-
-pub(crate) async fn write_to_device(
+pub(crate) async fn write_to_device<W>(
     image_path: PathBuf,
-    device_path: &PathBuf,
+    async_writer: &mut W,
     sender: &Sender<FlashProgress>,
-) -> Result<(u64, u64)> {
+) -> Result<(u64, u64)>
+where
+    W: AsyncWrite + std::marker::Unpin,
+{
     let img_file = fs::File::open(image_path).await?;
     let img_len = img_file.metadata().await?.len();
     let mut reader = io::BufReader::with_capacity(BUF_SIZE, img_file);
+    let mut writer = io::BufWriter::with_capacity(BUF_SIZE, async_writer);
 
-    let dev_file = fs::OpenOptions::new().write(true).open(device_path).await?;
-    let mut writer = io::BufWriter::with_capacity(BUF_SIZE, dev_file);
-
-    let mut buffer = vec![0; BUF_SIZE];
+    let mut buffer = [0u8; BUF_SIZE];
```

Comment:
This is 8 KiBs allocated on stack, which could overflow especially if BUF_SIZE is changed on a whim. Same for `calc_file_checksum()`

Reply:
`vec![]` is heap-allocated, an array is stack-allocated. Large arrays like these can overflow the stack, especially on our memory-limited system.

Reply:
Interesting, it seems the default stack size is set to 8kb. Thanks for making me aware of this

```
# ulimit -s
8192
```

Reply:
It's actually 8 MiB, but I'm still adamant about changes like these, since the original won't fail unexpectedly if we ever try changing the size of a buffer. There's even a (pedantic) [lint](https://rust-lang.github.io/rust-clippy/master/index.html#/large_stack_arrays) about this.